File: java/src/main/java/com/java/list/LinkedList.java
package com.java.list;

public class LinkedList<E> implements ListInterface<E> {
    private Node<E> head;
    private int numItems;

    public LinkedList() { //생성자
        numItems = 0;
        head = new Node<E>(null, null); //더미헤드
    }

    // main
    public static void main(String[] args) {
        LinkedList<String> list = new LinkedList<>();
        list.append("hello");
        list.append("java");
        list.append("world");

        for (int i = 0; i < list.len(); i++) {
            System.out.printf("%d번째: %s", i, list.get(i));
            System.out.println(" ");
        }

        list.add(1, "python");
        list.append("!");


        for (int i = 0; i < list.len(); i++) {
            System.out.printf("%d번째: %s", i, list.get(i));
            System.out.println(" ");
        }

        list.remove(0);
        list.add(0, "bye");

        for (int i = 0; i < list.len(); i++) {
            System.out.printf("%d번째: %s", i, list.get(i));
            System.out.println(" ");
        }

    }



    @Override
    // 연결 리스트에 원소 x 삽입하기
    public void add(int index, E item) {
        if (index >= 0 && index <= numItems) {
            Node<E> prevNode = getNode(index-1);
            Node<E> newNode = new Node<>(item, prevNode.next);
            prevNode.next = newNode;
            numItems++;
        }
    }

    @Override
    // 연결 리스트의 맨 뒤에 원소 추가하기
    public void append(E item) {
        Node<E> prevNode = head; //더미노드
        while (prevNode.next != null) {
            prevNode = prevNode.next; //next가 null일 때까지 찾으면 마지막 원소가 되니깐
        } // prevNode 준비끝
        Node<E> newNode = new Node<>(item, null);
        prevNode.next = newNode;
        numItems++;
    }

    @Override
    // 리스트 원소 삭제하기
    public E remove(int index) {
        if (index >=0 && index < numItems) {
            Node<E> prevNode = getNode(index - 1);
            Node<E> currNode = prevNode.next;
            prevNode.next = currNode.next;
            numItems--;
            return currNode.item;
        } else {
            return null;
        }
    }

    @Override
    public boolean removeItem(E x) {
        Node<E> prevNode, currNode = head;
        for (int i = 0; i <= numItems-1; i++) {
            prevNode = currNode;
            currNode = currNode.next;
//            if (currNode.item.compareTo(x) == 0) ;
            if (currNode.item == x) {
                prevNode.next = currNode.next;
                numItems--;
                return true;
            }
        }
        return false;
    }

    @Override
    // k번째 노드의 값 구하기
    public E get(int index) {
        if (index >= 0 && index <= numItems - 1) {
            return getNode(index).item;
        } else {
            return null;
        }
    }

    // k번째 노드 구하기
    public Node<E> getNode(int index) {
        if (index >= -1 && index <= numItems - 1) { // prevNode를 구할 ㄸㅐ -1을 해서 들어오니깐.
            Node<E> currNode = head;
            for (int i = 0; i <= index; i++) {
                currNode = currNode.next;
            }
            return currNode;
        } else {
            return null;
        }
    }

    @Override
    // 연결 리스트의 k번째 원소를 x로 대체하기
    public void set(int index, E x) {
        if (index >=0 && index <= numItems -1) {
            getNode(index).item = x;
        } else {
            /*예외처리*/
        }
    }

    public static final int NOT_FOUND = -1234;
    @Override
    public int indexOf(E x) {
        Node<E> currNode = head;
        int i;
        for (i = 0 ; i < numItems ; i++) {
            currNode = currNode.next;
            if (((Comparable)(currNode.item)).compareTo(x) == 0) {
                return i;
            }
        }
        return NOT_FOUND;
    }

    @Override
    public int len() {
        return numItems;
    }

    @Override
    public boolean isEmpty() {
        return numItems == 0;
    }

    @Override
    public void clear() {
        numItems = 0;
        head = new Node<>(null, null);

    }

}

File: kotlin/src/main/kotlin/com/kotlin/list/LinkedList.kt
package com.kotlin.list

class LinkedList : ListInterface {
    private var numItems: Int = 0
    private var head: Node
    private lateinit var prevNode: Node
    private lateinit var currNode: Node
    private lateinit var newNode: Node

    
    init {
        numItems = 0
        head = Node(null, null)
    }


    // 연결 리스트의 k번째 자리에 원소 x 삽입하기
    override fun add(index: Int, x: String) {
        if (index in 0..numItems) {
            prevNode = getNode(index-1)!!
            newNode = Node(x, prevNode.next)
            prevNode.next = newNode
            numItems++
        }
    }


    // 연결 리스트의 맨 뒤에 원소 추가하기
    override fun append(x: String) {
        prevNode = head
        while (prevNode.next != null) {
            prevNode = prevNode.next!!
        }

//            prevNode = getNode(numItems) // 이렇게 찾으면?
        newNode = Node(x, null)
        prevNode.next = newNode
        numItems++
    }


    // 연결 리스트의 k번째 원소 삭제하기
    override fun remove(index: Int): String? {
        if (index in 0 .. numItems) {
            prevNode = getNode(index-1)!!
            val removedNode = prevNode.next!!
            prevNode.next = prevNode.next!!.next
            numItems--
            return removedNode.item
            }
        return null

    }


    // 연결 리스트에서 원소 x 삭제하기
    override fun removeItem(x: String) : Boolean {
        prevNode = head
        whileloop@ while (prevNode.next != null) {
            currNode = prevNode.next!!
            if (currNode.item.equals(x)) break@whileloop
            prevNode = currNode
        }
        prevNode.next = prevNode.next!!.next
        numItems--
        return true
        
    }


    // k번째 노드의 값 구하기
    override fun get(index: Int) : String? {
        if (index in 0 until numItems) {
            return getNode(index)?.item
        }
        return null
    }
    
    private fun getNode(index: Int) : Node? {
        if (index in -1 until numItems) {
            currNode = head
            for (i in 0 .. index step 1) {
                currNode = currNode.next!!
            }
            return currNode
        }
        return null
    }
 

    // 배열 리스트의 i번째 원소를 x로 대체하기
    override fun set(index: Int, x: String) {
        getNode(index)?.item = x
    }


    // 원소x가 배열 리스트의 몇 번째 원소인지 알려주기
    private val NOT_FOUND = -12345
    override fun indexOf(x: String) : Int {
        currNode = head
        for (i in 0 until numItems step 1) {
            currNode = currNode.next!!
            if (currNode.item.equals(x)) return i
        }
        return NOT_FOUND
    }


    // 리스트의 총 원소 수 알려주기
    override fun len() : Int {
        return numItems
    }


    // 리스트가 비었는지 알려주기
    override fun isEmpty() : Boolean {
        return numItems == 0
    }


    // 리스트 깨끗이 청소하기
    override fun clear() {
        head = Node(null, null)
        numItems = 0
    }


}